internal/sync/cache.go

```diff
@@ -0,0 +1,371 @@
+
+	// The default limit of 50 is chosen based on experiments.
+	clusterCacheListSemaphoreSize int64 = 50
+
+	// clusterCacheListPageSize is the page size when performing K8s list requests.
+	// 500 is equal to kubectl's size
+	clusterCacheListPageSize int64 = 500
+
+	// clusterCacheListPageBufferSize is the number of pages to buffer when performing K8s list requests
+	clusterCacheListPageBufferSize int32 = 1
+
+	// clusterCacheRetryLimit sets a retry limit for failed requests during cluster cache sync
+	// If set to 1, retries are disabled.
+	clusterCacheAttemptLimit int32 = 1
+
+	// clusterCacheRetryUseBackoff specifies whether to use a backoff strategy on cluster cache sync, if retry is enabled
+	clusterCacheRetryUseBackoff bool = false
+)
+
+type ResourceInfo struct {
+	GitSyncName string
+
+	Health       *health.HealthStatus
+	manifestHash string
+}
+
+// LiveStateCache is a cluster caching that stores resource references and ownership
+// references. It also stored custom metadata for resources managed by GitSyncs. It always
+// ensures the cache is up-to-date before returning the resources.
+type LiveStateCache interface {
+	// GetClusterCache returns synced cluster cache
+	GetClusterCache() (clustercache.ClusterCache, error)
+	// GetManagedLiveObjs returns state of live nodes which correspond to target nodes of specified gitsync.
+	GetManagedLiveObjs(gitsync *v1alpha1.GitSync, targetObjs []*unstructured.Unstructured) (map[kube.ResourceKey]*unstructured.Unstructured, error)
+	// Init must be executed before cache can be used
+	Init()
+	// PopulateResourceInfo is called by the cache to update ResourceInfo struct for a managed resource
+	PopulateResourceInfo(un *unstructured.Unstructured, isRoot bool) (interface{}, bool)
+}
+
+type cacheSettings struct {
+	clusterSettings clustercache.Settings
+
+	// ignoreResourceUpdates is a flag to enable resource-ignore rules.
+	ignoreResourceUpdatesEnabled bool
+}
+
+type ObjectUpdatedHandler = func(managedByGitSync map[string]bool, ref v1.ObjectReference)
+
+type liveStateCache struct {
+	clusterCacheConfig *rest.Config
+	logger             *zap.SugaredLogger
+	// TODO: utilize `onObjectUpdated` when switch to change event triggering for sync tasking
+	// onObjectUpdated ObjectUpdatedHandler
+
+	cluster       clustercache.ClusterCache
+	cacheSettings cacheSettings
+	lock          sync.RWMutex
+}
+
+func newLiveStateCache(
+	cluster clustercache.ClusterCache,
+) LiveStateCache {
+	return &liveStateCache{
+		cluster: cluster,
+	}
+}
+
+func NewLiveStateCache(
+	clusterCacheConfig *rest.Config,
+) LiveStateCache {
+	return &liveStateCache{
+		clusterCacheConfig: clusterCacheConfig,
+	}
+}
+
+func (c *liveStateCache) getCluster() clustercache.ClusterCache {
+	c.lock.RLock()
+	if c.cluster != nil {
+		return c.cluster
+	}
+	cacheSettings := c.cacheSettings
+	c.lock.RUnlock()
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cluster != nil {
+		return c.cluster
+	}
```

Comment:
I guess the answer to my previous PR about "why do we re-check for this here?" is because it could've been nil when we checked above, but then set later prior to our entering our single-writer block, huh?

Reply:
yes.